ml-agents/mlagents/trainers/torch/attention.py

```diff
@@ -82,105 +101,85 @@ def forward(
-        entity_sizes: List[int],
+        entity_size: int,
+        entity_num_max_elements: Optional[int],
         embedding_size: int,
-        entity_num_max_elements: Optional[List[int]] = None,
         concat_self: bool = True,
     ):
         """
-        Constructs an EntityEmbeddings module.
+        Constructs an EntityEmbedding module.
         :param x_self_size: Size of "self" entity.
-        :param entity_sizes: List of sizes for other entities. Should be of length
-            equivalent to the number of entities.
-        :param embedding_size: Embedding size for entity encoders.
-        :param entity_num_max_elements: Maximum elements in an entity, None for unrestricted.
+        :param entity_size: Size of other entities.
+        :param entity_num_max_elements: Maximum elements for a given entity, None for unrestricted.
             Needs to be assigned in order for model to be exportable to ONNX and Barracuda.
-        :param concat_self: Whether to concatenate x_self to entites. Set True for ego-centric
+        :param embedding_size: Embedding size for the entity encoder.
+        :param concat_self: Whether to concatenate x_self to entities. Set True for ego-centric
             self-attention.
         """
         super().__init__()
         self.self_size: int = x_self_size
-        self.entity_sizes: List[int] = entity_sizes
-        self.entity_num_max_elements: List[int] = [-1] * len(entity_sizes)
+        self.entity_size: int = entity_size
+        self.entity_num_max_elements: int = -1
         if entity_num_max_elements is not None:
             self.entity_num_max_elements = entity_num_max_elements
-
-        self.concat_self: bool = concat_self
         # If not concatenating self, input to encoder is just entity size
         if not concat_self:
             self.self_size = 0
         # Initialization scheme from http://www.cs.toronto.edu/~mvolkovs/ICML2020_tfixup.pdf
-        self.ent_encoders = torch.nn.ModuleList(
-            [
-                LinearEncoder(
-                    self.self_size + ent_size,
-                    1,
-                    embedding_size,
-                    kernel_init=Initialization.Normal,
-                    kernel_gain=(0.125 / embedding_size) ** 0.5,
-                )
-                for ent_size in self.entity_sizes
-            ]
+        if self.self_size > 0:
+            self.self_encoder = LinearEncoder(
+                self.self_size,
+                1,
+                embedding_size // 2,
+                kernel_init=Initialization.Normal,
+                kernel_gain=(0.125 / embedding_size) ** 0.5,
+            )
+
+            self.ent_encoder = LinearEncoder(
+                self.entity_size,
+                1,
+                embedding_size - (embedding_size // 2),
+                kernel_init=Initialization.Normal,
+                kernel_gain=(0.125 / embedding_size) ** 0.5,
+            )
+
+        self.self_ent_encoder = LinearEncoder(
```

Comment:
This is an additional layer from my architecture. I was running with just ent_encoder.

Reply:
Experiments seem to indicate it works better with this extra embedding